magefile.publish.go

```diff
@@ -176,3 +177,97 @@ func PublishS3() error {
 
 	return nil
 }
+
+func walkMatch(root string, patterns []string) ([]string, error) {
+	var matches []string
+	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+
+		for _, pattern := range patterns {
+			if matched, err := filepath.Match(pattern, filepath.Base(path)); err != nil {
+				return err
+			} else if matched {
+				matches = append(matches, path)
+				return nil
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return matches, nil
+}
+
+func getPatterns(distro Distro) ([]string, error) {
+	// https://github.com/tarantool/tarantool/blob/c9e3926e5ae82785c5eb494521c44ba522109e95/tools/update_repo.sh#L986
+	if distro.OS == "el" || distro.OS == "fedora" {
+		return []string{"*.x86_64.rpm", "*.noarch.rpm"}, nil
+	}
+
+	// https://github.com/tarantool/tarantool/blob/c9e3926e5ae82785c5eb494521c44ba522109e95/tools/update_repo.sh#L517
+	// Modified to include only Linux tgz.
+	if distro.OS == "ubuntu" || distro.OS == "debian" {
+		return []string{"*.deb", "*.dsc", "*.Linux.*.tar.*z"}, nil
+	}
+
+	return nil, fmt.Errorf("Unknown OS: %s", distro.OS)
+}
+
+func PublishRWS() error {
+	fmt.Printf("Publish packages to RWS...\n")
+
+	for _, targetDistro := range targetDistros {
+		fmt.Printf("Publish package for %s/%s...\n", targetDistro.OS, targetDistro.Dist)
+
+		patterns, perr := getPatterns(targetDistro)
+		if perr != nil {
+			return fmt.Errorf("Failed to publish package for %s/%s: %s", targetDistro.OS, targetDistro.Dist, perr)
+		}
+
+		files, ferr := walkMatch(distPath, patterns)
+		if ferr != nil {
+			return fmt.Errorf("Failed to publish package for %s/%s: %s", targetDistro.OS, targetDistro.Dist, ferr)
+		}
+
+		rwsUrlPart := os.Getenv("RWS_URL_PART")
+		if rwsUrlPart == "" {
+			return fmt.Errorf("Failed to publish package: RWS_URL_PART is not set")
+		}
+
+		flags := []string{
+			"-v",
+			"-LfsS",
+			"-X", "PUT", fmt.Sprintf("%s/%s/%s", rwsUrlPart, targetDistro.OS, targetDistro.Dist),
+			"-F", fmt.Sprintf("product=%s", packageName),
+		}
+
+		for _, file := range files {
+			flags = append(flags, "-F", fmt.Sprintf("%s=@./%s", filepath.Base(file), file))
+		}
+
+		fmt.Printf("curl flags (excluding secrets): %s\n", flags)
+
+		rwsAuth := os.Getenv("RWS_AUTH")
+		if rwsAuth == "" {
+			return fmt.Errorf("Failed to publish package: RWS_AUTH is not set")
+		}
+		flags = append(flags, "-u", rwsAuth)
+
+		cmd := exec.Command("curl", flags...)
```

Comment:
Hm. Is not it is easier to use some HTTP client from Go?

Reply:
Easier? Not sure since there was already a prepared curl request in tarantool/queue. But I think it would be better in terms of error processing and general approach. But for now I don't know how to implement "-F" in Go http.

Reply:
Okay, let's ignore so.

Reply:
Well, seems like it's a bit complicated: https://github.com/mholt/curl-to-go/issues/34